core/src/main/scala/org/apache/spark/rpc/netty/NettyRpcEnv.scala

```diff
@@ -0,0 +1,485 @@
+    val promise = Promise[Any]()
+    val timeoutCancelable = nettyEnv.timeoutScheduler.schedule(new Runnable {
+      override def run(): Unit = {
+        promise.tryFailure(new TimeoutException("Cannot receive any reply in " + timeout.duration))
+      }
+    }, timeout.duration.toNanos, TimeUnit.NANOSECONDS)
+    val f = nettyEnv.ask(RequestMessage(nettyEnv.address, this, message, true))
+    f.onComplete { v =>
+      timeoutCancelable.cancel(true)
+      if (!promise.tryComplete(v)) {
+        logWarning(s"Ignore message $v")
+      }
+    }(ThreadUtils.sameThread)
+    promise.future.mapTo[T].recover(timeout.addMessageIfTimeout)(ThreadUtils.sameThread)
+  }
+
+  override def send(message: Any): Unit = {
+    require(message != null, "Message is null")
+    nettyEnv.send(RequestMessage(nettyEnv.address, this, message, false))
+  }
+
+  override def toString: String = s"NettyRpcEndpointRef(${_address})"
+
+  def toURI: URI = new URI(s"spark://${_address}")
+
+  final override def equals(that: Any): Boolean = that match {
+    case other: NettyRpcEndpointRef => _address == other._address
+    case _ => false
+  }
+
+  final override def hashCode(): Int = if (_address == null) 0 else _address.hashCode()
+}
+
+/**
+ * The message that is sent from the sender to the receiver.
+ */
+private[netty] case class RequestMessage(
+    senderAddress: RpcAddress, receiver: NettyRpcEndpointRef, content: Any, needReply: Boolean)
+
+/**
+ * The base trait for all messages that are sent back from the receiver to the sender.
+ */
+private[netty] trait ResponseMessage
+
+/**
+ * The reply for `ask` from the receiver side.
+ */
+private[netty] case class AskResponse(sender: NettyRpcEndpointRef, reply: Any)
+  extends ResponseMessage
+
+/**
+ * A message to send back to the receiver side. It's necessary because [[TransportClient]] only
+ * clean the resources when it receives a reply.
+ */
+private[netty] case class Ack(sender: NettyRpcEndpointRef) extends ResponseMessage
+
+/**
+ * A response that indicates some failure happens in the receiver side.
+ */
+private[netty] case class RpcFailure(e: Throwable)
+
+/**
+ * Maintain the mapping relations between client addresses and [[RpcEnv]] addresses, broadcast
+ * network events and forward messages to [[Dispatcher]].
+ */
+private[netty] class NettyRpcHandler(
+    dispatcher: Dispatcher, nettyEnv: NettyRpcEnv) extends RpcHandler with Logging {
+
+  private type ClientAddress = RpcAddress
+  private type RemoteEnvAddress = RpcAddress
+
+  // Store all client addresses and their NettyRpcEnv addresses.
+  @GuardedBy("this")
+  private val remoteAddresses = new mutable.HashMap[ClientAddress, RemoteEnvAddress]()
+
+  // Store the connections from other NettyRpcEnv addresses. We need to keep track of the connection
+  // count because `TransportClientFactory.createClient` will create multiple connections
+  // (at most `spark.shuffle.io.numConnectionsPerPeer` connections) and randomly select a connection
```

Comment:
It might be worth it to separate this config between shuffle and control rpcs; if someone increases the number of connections per peer for the shuffle backend, they may not necessarily want to do the same for the rpc backend.

This may apply to other rpc config options too, so perhaps we should track that as a separate enhancement.